Add docs for React Router App Server and adapters (#14144)
remix-run/react-router: docs/api/other-api/adapter.md

## `@react-router/express`

[Reference Documentation ↗](https://api.reactrouter.com/v7/modules/_react_router_express.html)

Here's an example with express:

```ts lines=[1-3,11-22]
const {
  createRequestHandler,
} = require("@react-router/express");
const express = require("express");

const app = express();

// needs to handle all verbs (GET, POST, etc.)
app.all(
  "*",
  createRequestHandler({
    // `react-router build` and `react-router dev` output files to a build directory,
    // you need to pass that build to the request handler
    build: require("./build"),

    // Return anything you want here to be available as `context` in your
    // loaders and actions. This is where you can bridge the gap between your
    // server and React Router
    getLoadContext(req, res) {
      return {};
    },
  }),
);
```

### Migrating from the React Router App Server

If you started an app with the [React Router App Server][rr-serve] but find that you want to take control over the Express server and customize it, it should be fairly straightforward to migrate way from `@react-router/serve`.

> You can refer to the [Express template][express-template] as a reference, gut here's the main changes you should have to make:

Reviewer: ```suggestion
You can refer to the [Express template][express-template] as a reference, but here are the main changes you will need to make:
```

---
title: "@react-router/{adapter}"
---

# Server Adapters

## Official Adapters

Idiomatic React Router apps can generally be deployed anywhere because React Router adapts the server's request/response to the [Web Fetch API][web-fetch-api]. It does this through adapters. We maintain a few adapters:

> - `@react-router/architect`

Reviewer: oof, I don't even want to mention this one 🙈

Almost wonder if it's worth also suggestion people checkout [SST](https://sst.dev/docs/start/aws/react/)

Though at that point, we probably have to do it for every third party adapter, which is not really the point here 🤔

Author: I think as long as we're publishing it we should probably include it :/.  If we want to straight up leave it out I think we'd be better off officially deprecating it or something.

And yeah linking to third parties is always a slippery slope to avoid any perceived favoritism...

Reviewer: Yeah, that's fine. But I do think we should consider deprecating it soon, or signal that we're going to deprecate it or something

[it seems to barely be used](https://npmtrends.com/@react-router/architect-vs-@remix-run/architect) btw

Author: ok, you convinced me lol - [0.3% of RR framework mode](https://npmtrends.com/@react-router/architect-vs-@react-router/cloudflare-vs-@react-router/dev) users are using it so maybe reference docs are enough for them

- `@react-router/express`

These adapters are imported into your server's entry and are not used inside your React Router app itself.

If you initialized your app with `npx create-react-router@latest` with something other than the built-in [React Router App Server][rr-serve] (`@react-router/serve`), you will note a `server/index.js` file that imports and uses one of these adapters.

<docs-info>If you're using the built-in React Router App Server, you don't interact with this API</docs-info>

Each adapter has the same API. In the future we may have helpers specific to the platform you're deploying to.

## `@react-router/express`

Here's a simple example with Express:

```tsx
import express from "express";

const BUILD_PATH = "./build/server/index.js";
const PORT = Number.parseInt(process.env.PORT || "3000");

const app = express();

app.use(
  "/assets",
  express.static("build/client/assets", {
    immutable: true,
    maxAge: "1y",
  }),
);
app.use(express.static("build/client", { maxAge: "1h" }));
app.use(await import(BUILD_PATH).then((mod) => mod.app));

app.listen(PORT, () => {
  console.log(
    `Server is running on http://localhost:${PORT}`,
  );
});
```

> ### Migrating from the React Router App Server

Author: This is a new section I added that we've long been missing

```tsx filename=vite.config.ts lines=[6-10]
import { reactRouter } from "@react-router/dev/vite";
import { defineConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";

export default defineConfig(({ isSsrBuild }) => ({
  build: {
    rollupOptions: isSsrBuild
      ? { input: "./server/app.ts" }
      : undefined,
  },
  plugins: [reactRouter(), tsconfigPaths()],
}));
```

**4. Update `package.json` scripts**

Update the `dev` and `start` scripts to use your new Express server:

```json filename=package.json
{
  // ...
  "scripts": {
    "dev": "cross-env NODE_ENV=development node server.js",
    "start": "node server.js"
    // ...
  }
  // ...
}
```

## `@react-router/architect`

Here's an example with Architect (AWS):

```ts
const {
  createRequestHandler,
} = require("@react-router/architect");
exports.handler = createRequestHandler({
  build: require("./build"),
});
```

## `@react-router/cloudflare`

Here's an example with the simplified Cloudflare Workers API:

```ts
import { createEventHandler } from "@react-router/cloudflare-workers";

import * as build from "../build";

addEventListener("fetch", createEventHandler({ build }));
```

> <!-- TODO: We used to have a Community Adapters section here, but unsure which of those are RR friendly so we should check that before re-including? -->

Author: @brookslybrand Curious what your thoughts are here and if we want to try to get something like https://remix.run/docs/en/main/other-api/adapter#community-adapters added back

---
title: "@react-router/{adapter}"
---

# Server Adapters

## Official Adapters

Idiomatic React Router apps can generally be deployed anywhere because React Router adapts the server's request/response to the [Web Fetch API][web-fetch-api]. It does this through adapters. We maintain a few adapters:

- `@react-router/architect`
- `@react-router/cloudflare`
- `@react-router/express`

These adapters are imported into your server's entry and are not used inside your React Router app itself.

If you initialized your app with `npx create-react-router@latest` with something other than the built-in [React Router App Server][rr-serve] (`@react-router/serve`), you will note a `server/index.js` file that imports and uses one of these adapters.

<docs-info>If you're using the built-in React Router App Server, you don't interact with this API</docs-info>

Each adapter has the same API. In the future we may have helpers specific to the platform you're deploying to.

## `@react-router/express`

> [Reference Documentation ↗](https://api.reactrouter.com/v7/modules/_react_router_express.html)

Reviewer: I think this example might be a bit confusing, it leaves out the part where you import `createRequestHandler` from `@react-router/express` and set it up

Author: oops - I must have copied in the wrong thing - the correct express example got left at the bottom of this doc by accident

Here's a simple example with Express:

```tsx
import express from "express";

const BUILD_PATH = "./build/server/index.js";
const PORT = Number.parseInt(process.env.PORT || "3000");

const app = express();

app.use(
  "/assets",
  express.static("build/client/assets", {
    immutable: true,
    maxAge: "1y",
  }),
);
app.use(express.static("build/client", { maxAge: "1h" }));
app.use(await import(BUILD_PATH).then((mod) => mod.app));

app.listen(PORT, () => {
  console.log(
    `Server is running on http://localhost:${PORT}`,
  );
});
```

### Migrating from the React Router App Server

> If you started an app with the [React Router App Server][rr-serve] but find that you want to take control over the Express server and customize it, it should be fairly straightforward to migrate way from `@react-router/serve`.

Reviewer: ```suggestion
You can refer to the [Express template][express-template] as a reference, but here are the main changes you should have to make:
```

Author: I'm confused lol - this sentence is below so this suggestion would duplicate it.  Do you just want to eliminate the "If you started an app with..." sentence?
